delegate-app/Main.hs

```diff
@@ -1,49 +1,179 @@
+  TChan,
+  TQueue,
+  atomically,
+  dupTChan,
+  flushTQueue,
+  newTChan,
+  newTQueue,
+  readTChan,
+  writeTChan,
+  writeTQueue,
+ )
+import Control.Monad (forever, (>=>))
+import Control.Monad.Trans (MonadIO (..), MonadTrans (lift))
+import Control.Tracer (Tracer, contramap, stdoutTracer)
+import Data.Aeson (ToJSON, eitherDecode, encode)
+import Data.Maybe (fromMaybe)
+import Network.WebSockets (
+  acceptRequest,
+  receiveData,
+  runServer,
+  sendTextData,
+  withPingThread,
+ )
+import Prettyprinter (Pretty (pretty))
 
-import Control.Concurrent (threadDelay)
-import Control.Monad.IO.Class (MonadIO (liftIO))
+-- Hydra imports
+import Hydra.Network (IP)
 
 -- HydraAuction imports
 import HydraAuction.Delegate (
   DelegateEvent (Start),
-  DelegateInput (DelegateEvent),
+  DelegateInput (DelegateEvent, FrontendRequest),
   DelegateRunnerT,
   delegateStep,
   execDelegateRunnerT,
  )
+import HydraAuction.Delegate.Interface (DelegateResponse, FrontendRequest)
+import HydraAuction.Delegate.Server (
+  DelegateError (FrontendNoParse),
+  DelegateServerConfig (
+    DelegateServerConfig,
+    dlgt'host,
+    dlgt'ping,
+    dlgt'port,
+    dlgt'tick
+  ),
+  DelegateServerLog (
+    DelegateError,
+    DelegateOutput,
+    FrontendConnected,
+    FrontendInput,
+    Started
+  ),
+  DelegateTracerT,
+  ServerAppT,
+ )
+import HydraAuction.Delegate.Tracing (MonadTracer (trace), TracerT, askTracer, runWithTracer')
+
+{- | per websocket, a thread gets spawned which then
+   subscribes to the broadcast channel and sends to
+   the incoming channel of the delegate runner
+-}
+delegateServerApp ::
+  -- | the amount of seconds between pings
+  Int ->
+  -- | the queue for the inputs of the delegates (this is where the requests are enqueued)
+  TQueue DelegateInput ->
+  -- | the queue for the broadcast (this is where we receives events from the runner)
+  TChan DelegateResponse ->
+  ServerAppT (DelegateTracerT IO)
+delegateServerApp pingSecs reqq broadcast pending = do
+  connection <- liftIO $ acceptRequest pending
+
+  trace FrontendConnected
+  tracer <- askTracer
+
+  liftIO $ do
+    broadcast' <- atomically $ dupTChan broadcast
+
+    let encodeSend :: forall a. ToJSON a => a -> IO ()
+        encodeSend = sendTextData connection . encode
+
+        receive = forever $
+          runWithTracer' tracer $ do
+            inp <- liftIO $ receiveData connection
+            case eitherDecode @FrontendRequest inp of
+              Left msg -> do
+                let err = FrontendNoParse msg
+                trace (DelegateError err)
+                liftIO $ encodeSend err
+              Right req -> do
+                trace $ FrontendInput req
+                liftIO . atomically . writeTQueue reqq $ FrontendRequest req
+
+        send =
+          forever $
+            atomically (readTChan broadcast')
+              >>= encodeSend
+
+    withPingThread connection pingSecs (pure ()) $ race_ receive send -- TODO: can we do this
```

Comment:
-- Many (but not all) proxies have a 60 second default timeout, so based on that
-- sending a ping every 30 seconds is a good idea.

Please add comment on why code uses ping.